qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -160,83 +180,198 @@ def __init__(
 
     def run(self, dag: DAGCircuit) -> DAGCircuit:
         """Run the HighLevelSynthesis pass on `dag`.
+
         Args:
             dag: input dag.
+
         Returns:
-            Output dag with certain Operations synthesized (as specified by
-            the hls_config).
+            Output dag with higher-level operations synthesized.
 
         Raises:
-            TranspilerError: when the specified synthesis method is not available.
+            TranspilerError: when the transpiler is unable to synthesize the given DAG
+            (for instance, when the specified synthesis method is not available).
         """
-        # If there aren't any high level operations to synthesize return fast
+
+        # In the future this pass will be recursive as we may have annotated gates
+        # whose definitions consist of other annotated gates, whose definitions include
+        # LinearFunctions. Note that in order to synthesize a controlled linear
+        # function, we must first fully synthesize the linear function, and then
+        # synthesize the circuit obtained by adding control logic.
+        # Additionally, see https://github.com/Qiskit/qiskit/pull/9846#pullrequestreview-1626991425.
+
+        # If there are no high level operations / annotated gates to synthesize, return fast
         hls_names = set(self.hls_plugin_manager.plugins_by_op)
-        if not hls_names.intersection(dag.count_ops()):
+        node_names = dag.count_ops()
+        if "annotated" not in node_names and not hls_names.intersection(node_names):
             return dag
-        for node in dag.op_nodes():
-            if node.name in self.hls_config.methods.keys():
-                # the operation's name appears in the user-provided config,
-                # we use the list of methods provided by the user
-                methods = self.hls_config.methods[node.name]
-            elif (
-                self.hls_config.use_default_on_unspecified
-                and "default" in self.hls_plugin_manager.method_names(node.name)
-            ):
-                # the operation's name does not appear in the user-specified config,
-                # we use the "default" method when instructed to do so and the "default"
-                # method is available
-                methods = ["default"]
+
+        # copy dag_op_nodes because we are modifying the DAG below
+        dag_op_nodes = dag.op_nodes()
+
+        for node in dag_op_nodes:
+            qubits = (
+                [dag.find_bit(x).index for x in node.qargs] if self._use_qubit_indices else None
+            )
+            decomposition = self._recursively_handle_op(node.op, qubits)
+
+            if not isinstance(decomposition, (QuantumCircuit, Operation)):
+                raise TranspilerError(f"HighLevelSynthesis was unable to synthesize {node.op}.")
+
+            if isinstance(decomposition, QuantumCircuit):
```

Comment:
This is making me think in parallel we should have an option for a plugin to return `DAGCircuits` to avoid this conversion.

Reply:
I was thinking about this as well! Related to this, in the follow-up PR #10965 I have already allowed some of the internal functions like `_recursively_handle_op` to also return a `DAGCircuit`, otherwise there are too many useless back-and-forth conversions.